File: 733-flood-fill/flood-fill.py
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        row = len(image)
        col = len(image[0])
        r = image[sr][sc]
        if(r == color):
            return image
        def dfs(sr,sc,color):
            if(sr<0 or sc<0 or sr>= row or sc>= col):
                return
            if(image[sr][sc]!= r):
                return
            image[sr][sc] = color
            dfs(sr+1, sc, color)
            dfs(sr-1, sc, color)
            dfs(sr, sc+1, color)
            dfs(sr, sc-1, color)
        dfs(sr, sc, color)
        return image

File: 733-flood-fill/flood-fill.cpp
class Solution {
private:
    void dfs(int sr, int sc, int color, vector<vector<int>>& image, vector<vector<int>>&res, int initial, int drow[], int dcol[]){
        res[sr][sc] = color;
        int n = image.size();
        int m = image[0].size();
        for(int i =0;i<4;i++){
            int nrow = drow[i] + sr;
            int ncol = dcol[i] + sc;
            if(nrow>=0 && nrow<n && ncol>=0 && ncol<m && image[nrow][ncol]== initial && res[nrow][ncol]!=color ){
                dfs(nrow, ncol,color, image, res, initial, drow, dcol);
            }
        }
    }
public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
        int initial = image[sr][sc];
        vector<vector<int>> res = image;
        int drow[] = {-1, 0, 1, 0};
        int dcol[] = {0, 1, 0, -1};
        dfs(sr, sc, color, image, res,initial, drow, dcol);
        return res; 
    }
};